### PR TITLE
Add support for using cmake in the mono runtime build.

### PR DESCRIPTION
**Pull request by vargaz**
Not enabled by default. Ninja is used for building if available, else we fall
back to make.

### CODE REVIEW

File: src/mono/mono.proj [hunk 1]
```diff
@@ -933,16 +938,49 @@
     <Error Condition="!Exists('$(LocalDotnet)')" Text="'$(LocalDotnet)' doesn't exist." />
   </Target>
 
+  <Target Name="BuildMonoRuntimeCMake" Condition="'$(MonoEnableCMake)' != ''">
```
Comment (akoeplinger) [inline src/mono/mono.proj:941]: better to use this so `MonoEnableCMake=false` doesn't match:
```suggestion
  <Target Name="BuildMonoRuntimeCMake" Condition="'$(MonoEnableCMake)' == 'true'">
```

File: src/mono/mono.proj [hunk 2]
```diff
@@ -755,7 +760,7 @@
     <Touch Condition="'$(_MonoAotCrossConfigureCommand)' != ''" Files="$(MonoObjDir)cross/config.h" />
   </Target>
 
-  <Target Name="BuildMonoRuntimeUnix" Condition="'$(OS)' != 'Windows_NT'" DependsOnTargets="ConfigureMonoRuntimeUnix">
+  <Target Name="BuildMonoRuntimeUnix" Condition="'$(OS)' != 'Windows_NT' and '$(MonoEnableCMake)' == ''" DependsOnTargets="ConfigureMonoRuntimeUnix">
```
Comment (akoeplinger) [inline src/mono/mono.proj:763]: better to use this so `MonoEnableCMake=false` doesn't match:
```suggestion
  <Target Name="BuildMonoRuntimeUnix" Condition="'$(OS)' != 'Windows_NT' and '$(MonoEnableCMake)' != 'true'" DependsOnTargets="ConfigureMonoRuntimeUnix">
```

File: src/mono/mono.proj [hunk 3]
```diff
@@ -933,16 +938,49 @@
...
+  <Target Name="BuildMonoRuntimeCMake" Condition="'$(MonoEnableCMake)' != ''">
+    <Exec Command="which ninja" IgnoreExitCode="true" IgnoreStandardErrorWarningFormat="true">
+      <Output TaskParameter="ExitCode" PropertyName="MonoNinjaFound"/>
+    </Exec>
+    <PropertyGroup>
+      <MonoBuildTool Condition="'$(MonoNinjaFound)' == '0'">ninja</MonoBuildTool>
+      <MonoBuildTool Condition="'$(MonoNinjaFound)' != '0'">make -j$([System.Environment]::ProcessorCount)</MonoBuildTool>
+    </PropertyGroup>
+    <ItemGroup>
+      <MonoCMakeArgs Condition="'$(MonoNinjaFound)' == '0'" Include="-G Ninja"/>
```
Comment (akoeplinger) [inline src/mono/mono.proj:950]: nit:
```suggestion
      <MonoCMakeArgs Condition="'$(MonoBuildTool)' == 'ninja'" Include="-G Ninja"/>
```

File: src/mono/mono.proj [hunk 4]
```diff
@@ -933,16 +938,49 @@
...
+      <MonoCMakeArgs Include="-DENABLE_ICALL_EXPORT=1"/>
+      <MonoCMakeArgs Include="-DENABLE_LAZY_GC_THREAD_CREATION=1"/>
+      <MonoCMakeArgs Include="-DCMAKE_C_FLAGS=-I$(PkgMicrosoft_NETCore_Runtime_ICU_Transport)/runtimes/browser-wasm/native/include"/>
+    </ItemGroup>
+    <MakeDir Directories="$(MonoObjDir)" />
+    <Message Text="Running cmake with arguments: @(MonoCMakeArgs, ' ') $(MonoProjectRoot)" Importance="High"/>
+    <!-- cmake -->
+    <Exec Condition="'$(TargetsBrowser)' != 'true'" Command="cmake @(MonoCMakeArgs, ' ') $(MonoProjectRoot)" IgnoreStandardErrorWarningFormat="true" WorkingDirectory="$(MonoObjDir)"/>
+    <Exec Condition="'$(TargetsBrowser)' == 'true'" Command="bash -c 'source $(EMSDK_PATH)/emsdk_env.sh &amp;&amp; emcmake cmake @(MonoCMakeArgs, ' ') $(MonoProjectRoot)'" IgnoreStandardErrorWarningFormat="true" WorkingDirectory="$(MonoObjDir)" />
+    <!-- ninja -->
```
Comment (akoeplinger) [inline src/mono/mono.proj:967]:
```suggestion
    <!-- ninja/make -->
```

File: src/mono/mono.proj [hunk 5]
```diff
@@ -1004,4 +1042,8 @@
     <Exec Condition="'$(TargetsOSX)' == 'true' or '$(TargetsiOS)' == 'true' or '$(TargetstvOS)' == 'true'" Command="install_name_tool -id @rpath/$(MonoFileName) $(BinDir)$(MonoFileName)" />
   </Target>
 
+  <Target Name="CleanMono">
+    <RemoveDir Directories="$(MonoObjDir)"/>
```
Comment (akoeplinger) [inline src/mono/mono.proj:1046]:
```suggestion
    <RemoveDir Directories="$(MonoObjDir)" />
```

**Review by akoeplinger — COMMENTED** [review]
A couple nits, looks great otherwise!

**Comment by vargaz** [comment]
Updated.

File: src/mono/mono.proj [hunk 6]
```diff
@@ -755,7 +760,7 @@
     <Touch Condition="'$(_MonoAotCrossConfigureCommand)' != ''" Files="$(MonoObjDir)cross/config.h" />
   </Target>
 
-  <Target Name="BuildMonoRuntimeUnix" Condition="'$(OS)' != 'Windows_NT'" DependsOnTargets="ConfigureMonoRuntimeUnix">
+  <Target Name="BuildMonoRuntimeUnix" Condition="'$(OS)' != 'Windows_NT' and '$(MonoEnableCMake)' != 'true'" DependsOnTargets="ConfigureMonoRuntimeUnix">
```
Comment (lateralusX) [inline src/mono/mono.proj:763]: Just to prepare for potential future use, should we add the same condition to the Windows build target?

File: src/mono/mono.proj [hunk 7]
```diff
@@ -933,16 +938,52 @@
     <Error Condition="!Exists('$(LocalDotnet)')" Text="'$(LocalDotnet)' doesn't exist." />
   </Target>
 
+  <Target Name="BuildMonoRuntimeCMake" Condition="'$(MonoEnableCMake)' == 'true'">
```
Comment (lateralusX) [inline src/mono/mono.proj:941]: Looks specific to a couple of platform builds at the moment (osx, browser), should we have an error if you try to enable cmake build for a platform that is currently not supported?

File: src/mono/mono.proj [hunk 8]
```diff
@@ -933,16 +938,52 @@
...
+    <PropertyGroup>
+      <MonoBuildTool Condition="'$(MonoNinjaFound)' == '0'">ninja</MonoBuildTool>
+      <MonoBuildTool Condition="'$(MonoNinjaFound)' != '0'">make -j$([System.Environment]::ProcessorCount)</MonoBuildTool>
+    </PropertyGroup>
+    <ItemGroup>
+      <MonoCMakeArgs Condition="'$(MonoBuildTool)' == 'ninja'" Include="-G Ninja"/>
+      <MonoCMakeArgs Include="-DCMAKE_INSTALL_PREFIX=$(MonoObjDir)out"/>
+    </ItemGroup>
+    <ItemGroup Condition="'$(TargetsOSX)' == 'true'">
+      <MonoCMakeArgs Include="-DCMAKE_C_FLAGS=-mmacosx-version-min=10.12"/>
```
Comment (akoeplinger) [inline src/mono/mono.proj:954]: we already have a property for the min OSX version we support:
```suggestion
      <MonoCMakeArgs Include="-DCMAKE_C_FLAGS=-mmacosx-version-min=$(macOSVersionMin)"/>
```

that said, it looks like we're missing this for the autotools-based version too. I'll send a PR to fix that.

Comment (akoeplinger) [reply src/mono/mono.proj:954]: alternatively I think the preferred way to set this in CMake is to use `-DCMAKE_OSX_DEPLOYMENT_TARGET=10.13`

**Comment by vargaz** [comment]
Currently, the debug/release builds are the same because we don't pick up the MonoCFlags property since its set inside the
ConfigureMonoRuntimeUnix target which is not used by the cmake build since it doesn't have a separate configure pass.

File: src/mono/mono.proj (hunk 7 above)
Comment (vargaz) [reply src/mono/mono.proj:941]: It's not supported on any platforms right now.

**Comment by akoeplinger** [comment]
> because we don't pick up the MonoCFlags property

I _think_ that's fine since ideally we should set these flags inside of the CMakeLists.txt and not in mono.proj eventually.

**Review by akoeplinger — APPROVED** [review]

**Pull request merged**